x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java

```diff
@@ -0,0 +1,65 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical;
+
+import org.elasticsearch.xpack.esql.core.expression.Alias;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.AggregateFunction;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.Count;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.CountDistinct;
+import org.elasticsearch.xpack.esql.plan.logical.Aggregate;
+import org.elasticsearch.xpack.esql.plan.logical.Eval;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.Project;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Replaces an aggregation function having a false/null filter with an EVAL node.
+ * <pre>
+ *     ... | STATS x = someAgg(y) WHERE FALSE {BY z} | ...
+ *     =>
+ *     ... | STATS x = someAgg(y) {BY z} > | EVAL x = NULL | KEEP x{, z} | ...
+ * </pre>
+ */
+public class ReplaceStatsFilteredAggWithEval extends OptimizerRules.OptimizerRule<Aggregate> {
+    @Override
+    protected LogicalPlan rule(Aggregate aggregate) {
+        List<NamedExpression> newAggs = new ArrayList<>(aggregate.aggregates().size());
+        List<Alias> newEvals = new ArrayList<>(aggregate.aggregates().size());
+        List<NamedExpression> newProjections = new ArrayList<>(aggregate.aggregates().size());
+        for (var ne : aggregate.aggregates()) {
+            if (ne instanceof Alias alias
+                && alias.child() instanceof AggregateFunction aggFunction
+                && aggFunction.hasFilter()
+                && aggFunction.filter() instanceof Literal literal
+                && literal.fold().equals(false)) {
+
+                Object value = aggFunction instanceof Count || aggFunction instanceof CountDistinct ? 0L : null;
+                Alias newAlias = alias.replaceChild(Literal.of(aggFunction, value));
+                newEvals.add(newAlias);
+                newProjections.add(newAlias.toAttribute());
+            } else {
+                newAggs.add(ne); // agg function unchanged or grouping key
+                newProjections.add(ne.toAttribute());
+            }
+        }
+
+        LogicalPlan plan = aggregate;
+        if (newEvals.isEmpty() == false) {
+            plan = newEvals.size() < aggregate.aggregates().size() - aggregate.groupings().size()
+                ? aggregate.with(aggregate.child(), aggregate.groupings(), newAggs)
+                : aggregate.child();
+            plan = new Eval(aggregate.source(), plan, newEvals);
+            plan = new Project(aggregate.source(), plan, newProjections);
```

Comment:
I believe the eval can be skipped and instead just use the project with aliases to the underlying literal.
Can you confirm that CombineProjection eliminates the underlying aggregation (count, max, etc...)?
e.g.:
```
FROM x | STATS c = COUNT() WHERE FALSE BY y  
// becomes
FROM x | STATS c = COUNT() WHERE FALSE by y | EVAL c = 0
// which is optimized to
FROM x |  STATS BY y | EVAL c= 0
```

Reply:
> I believe the eval can be skipped and instead just use the project with aliases to the underlying literal.

Not sure if this is possible without updating the plan verification: `PlanConsistencyChecker#checkPlan()` would no longer find the references, since `Alias` would delegate to `Literal`, which returns empty refs. Also, would this not prevent some other rules to apply, like `PropagateEvalFoldables`?

> Can you confirm that CombineProjection eliminates the underlying aggregation (count, max, etc...)?

Yes, the underling aggregation is being replaced ([test here](https://github.com/elastic/elasticsearch/pull/115858/files#diff-427544eaca2f06adb1b5850278a22573fb696564260a34a8f5580ff66bbbd95aR821)).